## Feature Analysis
This document describes the features that were analyzed for the SH2 domain as reported in this directory
These reports were generated by the CoDIAC package on: 07/17/2024 10:21:09
### Analysis Reports (CSV files)
Features were analyzed with regard to:
1. The proximity to a conserved ligand binding sites
2. The direct intersection with domain-domain interactions
3. The number of other features of that type in the alignment
### Feature Cutoffs and files used
Ligand binding was considered conserved if >= 8 residues in that alignment position were in the ligand binding features
The following files were used:
1. Data/Uniprot_Reference/alignment/SH2_IPR000980_promals3d.fasta
2. Data/Features/interDomain/SH2_IPR000980_interDomain_phospholipid.feature
3. Data/Features/ligand/SH2_IPR000980_Ligand.feature
4. PTMs in the following directory Data/Features/PTMS_all/PTMs_final/
### Analysis Report Columns
1. header: the fasta header of the sequence
2. Uniprot_ID: the uniprot ID of the sequence
3. gene_name: the gene name of the sequence
4. SH2 number: the SH2 domain number (1 if N-terminal and 2 if C-terminal domain in 2 SH2 domain proteins
5. protein_residue: the residue in the protein sequence numbering
6. feature_position: the position in the domain, as printed in the feature file, relative to SH2 start
7. alignment_postion: the position in the alignment
8. number_mods_at_position: the number of modifications of this type at that position
9. number_conserved_ligand_residues_nearby: the number of residues in the ligand binding site at that position within +/- 2 amino acids away
10. number_residues_from_conserved_ligand: the number of residues from the conserved ligand binding site. If 0, it falls directly on a ligand interface.
11. domain_interaction: the domain interaction at that position, if it intersects directly with a domain-domain interface. We also use this column to indicate if it interacts with a residue identified as interacting with phospholipids
### PTM Analysis
The following PTMs were analyzed:
1. N6-acetyllysine
2. Phosphoserine
3. Phosphothreonine
4. Phosphotyrosine
5. Ubiquitination
6. Phosphorylation (Phosphoserine and Phosphothreonine combined)
